tests/transactions/tests.py

```diff
@@ -498,3 +498,69 @@ def test_orm_query_without_autocommit(self):
         finally:
             transaction.rollback()
             transaction.set_autocommit(True)
+
+
+class DurableTests(TransactionTestCase):
+    '''Tests for the durable flag to atomic.'''
+    available_apps = ['transactions']
+
+    def test_commit(self):
+        with transaction.atomic(durable=True):
+            reporter = Reporter.objects.create(first_name="Tintin")
+        self.assertSequenceEqual(Reporter.objects.all(), [reporter])
+
+    def test_nested_outer_durable(self):
+        with transaction.atomic(durable=True):
+            reporter1 = Reporter.objects.create(first_name="Tintin")
+            with transaction.atomic():
+                reporter2 = Reporter.objects.create(first_name="Archibald", last_name="Haddock")
+        self.assertSequenceEqual(Reporter.objects.all(), [reporter2, reporter1])
+
+    def test_nested_both_durable(self):
+        message = 'A durable atomic block cannot be nested within another atomic block.'
+        with transaction.atomic(durable=True):
+            with self.assertRaisesMessage(RuntimeError, message):
+                with transaction.atomic(durable=True):
+                    pass
+
+    def test_nested_inner_durable(self):
+        message = 'A durable atomic block cannot be nested within another atomic block.'
+        with transaction.atomic():
+            with self.assertRaisesMessage(RuntimeError, message):
+                with transaction.atomic(durable=True):
+                    pass
+
+
+class DisableDurablitityCheckTests(TestCase):
+    '''
+    Tests showing the disabling of durablility checking in TestCase.
+
+    Because TestCase runs all tests in a transaction by default, code using
+    durable=True would always fail when run from TestCase. This would mean
+    these tests would be forced to use the slower TransactionTestCase even
+    when not testing durability. For this reason, TestCase disables the
+    durability check.
+    '''
+    available_apps = ['transactions']
+
+    def test_commit(self):
+        with transaction.atomic(durable=True):
+            reporter = Reporter.objects.create(first_name="Tintin")
+        self.assertSequenceEqual(Reporter.objects.all(), [reporter])
+
+    def test_nested_outer_durable(self):
+        with transaction.atomic(durable=True):
+            reporter1 = Reporter.objects.create(first_name="Tintin")
+            with transaction.atomic():
+                reporter2 = Reporter.objects.create(first_name="Archibald", last_name="Haddock")
+        self.assertSequenceEqual(Reporter.objects.all(), [reporter2, reporter1])
+
+    def test_nested_both_durable(self):
+        with transaction.atomic(durable=True):
+            with transaction.atomic(durable=True):
+                pass
+
+    def test_nested_inner_durable(self):
+        with transaction.atomic():
+            with transaction.atomic(durable=True):
+                pass
```

Comment:
I think we should raise an error if someone is trying to use `durable` in the `TestCase` class, instead of ignoring it, e.g.
```python
if self.durable and not self.ensure_durability:
    raise ....
```

Reply:
What do you mean here @felixxm ? Do you mean to add the ability for `Atomic` to discern between the two levels of `@atomic` created by `TestCase` and the one inside the test, so that this case can raise an error? IMO that sounds too complicated to be worth it.

Reply:
This flag is ignored when `ensure_durability` is `False`, so we should inform users that is not allowed, e.g.
```diff
(django-test) git:pr/13708 felixx@felixx-A555:~/repo/django/tests> git diff
diff --git a/django/db/transaction.py b/django/db/transaction.py
index c6ba346a99..8a84b97237 100644
--- a/django/db/transaction.py
+++ b/django/db/transaction.py
@@ -172,6 +172,11 @@ class Atomic(ContextDecorator):
         self.using = using
         self.savepoint = savepoint
         self.durable = durable
+        if self.durable and not self.ensure_durability:
+            raise ValueError(
+                'A durable atomic block is not allowed. If you are running '
+                'tests, you must use TransactionTestCase, not TestCase.'
+            )
 
     def __enter__(self):
         connection = get_connection(self.using)
```
We can be descriptive here.

Reply:
From what I understand, the point of being able to turn off durability checking (via `ensure_durability`) is that durable atomic blocks _will_ be able to be run within a `TestCase`. So I think it's correct that it should ignore, rather than error.

Reply:
I think this would be a massive usability loss - I want people to be able to continue testing most of the features of their code in `TestCase` even if it happens to set the `durable` flag. For testing durability itself, `TransactionTestCase` is appropriate of course.

Reply:
@seddonym is right - if we don't want to support running code using `durable` in `TestCase`, we can just remove `ensure_durability` and allow the `RuntimeError`.

Reply:
Right :+1: OK let's leave it as it is.